File: src/main/java/com/dh/ondot/schedule/domain/service/ScheduleService.java
package com.dh.ondot.schedule.domain.service;

import com.dh.ondot.member.domain.Member;
import com.dh.ondot.schedule.core.exception.NotFoundScheduleException;
import com.dh.ondot.schedule.domain.Schedule;
import com.dh.ondot.schedule.domain.repository.ScheduleRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class ScheduleService {
    private final ScheduleRepository scheduleRepository;

    public Schedule setupSchedule(
            Member member, LocalDateTime appointmentAt, int estimatedTimeMin
    ) {
        Optional<Schedule> schedule = scheduleRepository.findFirstByMemberIdOrderByUpdatedAtDesc(member.getId());

        Schedule newSchedule;
        if (schedule.isPresent()) {
            newSchedule = createFromLatestUserSetting(schedule.get(), member, appointmentAt, estimatedTimeMin);
        } else {
            newSchedule = Schedule.createWithDefaultAlarmSetting(
                    member.getDefaultAlarmMode(), member.getSnooze(), member.getSound(),
                    appointmentAt, estimatedTimeMin, member.getPreparationTime()
            );
        }
        newSchedule.setupQuickSchedule(member.getId(), appointmentAt);

        return newSchedule;
    }

    private Schedule createFromLatestUserSetting(
            Schedule latestSchedule, Member member,
            LocalDateTime appointment, int estimatedTimeMin
    ) {
        Schedule copy = copySchedule(latestSchedule);

        LocalDateTime depAlarmAt = appointment.minusMinutes(estimatedTimeMin);
        LocalDateTime prepAlarmAt = depAlarmAt.minusMinutes(member.getPreparationTime());

        copy.getDepartureAlarm().updateTriggeredAt(depAlarmAt);
        copy.getPreparationAlarm().updateTriggeredAt(prepAlarmAt);

        return copy;
    }

    @Transactional
    public Schedule saveSchedule(Schedule schedule) {
        return scheduleRepository.save(schedule);
    }

    public Schedule findScheduleById(Long id) {
        return scheduleRepository.findById(id)
                .orElseThrow(() -> new NotFoundScheduleException(id));
    }

    @Transactional
    public void deleteSchedule(Schedule schedule) {
        scheduleRepository.delete(schedule);
    }

    private Schedule copySchedule(Schedule original) {
        return Schedule.builder()
                .preparationAlarm(original.getPreparationAlarm().copy())
                .departureAlarm(original.getDepartureAlarm().copy())
                .build();
    }
}

File: src/main/kotlin/com/dh/ondot/schedule/domain/service/ScheduleService.kt
package com.dh.ondot.schedule.domain.service

import com.dh.ondot.member.domain.Member
import com.dh.ondot.schedule.domain.Place
import com.dh.ondot.schedule.domain.Schedule
import com.dh.ondot.schedule.domain.enums.TransportType
import com.dh.ondot.schedule.domain.repository.ScheduleRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Instant
import java.time.LocalDateTime
import java.util.SortedSet

@Service
class ScheduleService(
    private val scheduleRepository: ScheduleRepository,
) {
    fun createScheduleWithAlarms(
        member: Member, appointmentAt: LocalDateTime, estimatedTimeMin: Int,
    ): Schedule {
        val schedule = scheduleRepository.findFirstByMemberIdOrderByUpdatedAtDesc(member.id)

        return if (schedule.isPresent) {
            createFromLatestUserSetting(schedule.get(), member, appointmentAt, estimatedTimeMin)
        } else {
            Schedule.createWithDefaultAlarmSetting(
                member.defaultAlarmMode!!, member.snooze!!, member.sound!!,
                appointmentAt, estimatedTimeMin, member.preparationTime!!,
            )
        }
    }

    fun setupSchedule(
        member: Member, appointmentAt: LocalDateTime, estimatedTimeMin: Int,
    ): Schedule {
        val newSchedule = createScheduleWithAlarms(member, appointmentAt, estimatedTimeMin)
        newSchedule.setupQuickSchedule(member.id, appointmentAt)
        return newSchedule
    }

    fun createEverytimeSchedule(
        member: Member, departurePlace: Place, arrivalPlace: Place,
        title: String, repeatDays: SortedSet<Int>, appointmentAt: LocalDateTime,
        estimatedTimeMin: Int, transportType: TransportType,
    ): Schedule {
        val base = createScheduleWithAlarms(member, appointmentAt, estimatedTimeMin)
        return saveSchedule(Schedule.createSchedule(
            memberId = member.id,
            departurePlace = departurePlace,
            arrivalPlace = arrivalPlace,
            preparationAlarm = base.preparationAlarm!!,
            departureAlarm = base.departureAlarm!!,
            title = title,
            isRepeat = true,
            repeatDays = repeatDays,
            appointmentAt = appointmentAt,
            transportType = transportType,
        ))
    }

    private fun createFromLatestUserSetting(
        latestSchedule: Schedule, member: Member,
        appointment: LocalDateTime, estimatedTimeMin: Int,
    ): Schedule {
        val copy = copySchedule(latestSchedule)
        copy.preparationAlarm!!.changeEnabled(true)
        copy.departureAlarm!!.changeEnabled(true)
        val depAlarmAt = appointment.minusMinutes(estimatedTimeMin.toLong())
        val prepAlarmAt = depAlarmAt.minusMinutes(member.preparationTime!!.toLong())
        copy.departureAlarm!!.updateTriggeredAt(depAlarmAt)
        copy.preparationAlarm!!.updateTriggeredAt(prepAlarmAt)

        return copy
    }

    private fun copySchedule(original: Schedule): Schedule =
        Schedule(
            preparationAlarm = original.preparationAlarm!!.copy(),
            departureAlarm = original.departureAlarm!!.copy(),
        )

    fun getEarliestActiveAlarmAt(schedules: List<Schedule>): Instant? =
        schedules
            .filter { it.hasAnyActiveAlarm() }
            .mapNotNull { it.calculateNextAlarmAt() }
            .minOrNull()

    @Transactional
    fun saveSchedule(schedule: Schedule): Schedule =
        scheduleRepository.save(schedule)

    @Transactional
    fun deleteSchedule(schedule: Schedule) {
        scheduleRepository.delete(schedule)
    }

    @Transactional
    fun deleteAllByMemberId(memberId: Long) {
        scheduleRepository.deleteByMemberId(memberId)
    }
}
